crates/nu-protocol/src/pipeline/byte_stream.rs

```diff
@@ -88,76 +137,162 @@ impl Read for SourceReader {
 ///
 /// Internally, [`ByteStream`]s currently come in three flavors according to [`ByteStreamSource`].
 /// See its documentation for more information.
+#[derive(Debug)]
 pub struct ByteStream {
     stream: ByteStreamSource,
     span: Span,
     ctrlc: Option<Arc<AtomicBool>>,
+    type_: ByteStreamType,
     known_size: Option<u64>,
 }
 
 impl ByteStream {
     /// Create a new [`ByteStream`] from a [`ByteStreamSource`].
-    pub fn new(stream: ByteStreamSource, span: Span, interrupt: Option<Arc<AtomicBool>>) -> Self {
+    pub fn new(
+        stream: ByteStreamSource,
+        span: Span,
+        interrupt: Option<Arc<AtomicBool>>,
+        type_: ByteStreamType,
+    ) -> Self {
         Self {
             stream,
             span,
             ctrlc: interrupt,
+            type_,
             known_size: None,
         }
     }
 
-    /// Create a new [`ByteStream`] from a [`ByteStreamSource::Read`].
+    /// Create a [`ByteStream`] from an arbitrary reader. The type must be provided.
     pub fn read(
         reader: impl Read + Send + 'static,
         span: Span,
         interrupt: Option<Arc<AtomicBool>>,
+        type_: ByteStreamType,
     ) -> Self {
-        Self::new(ByteStreamSource::Read(Box::new(reader)), span, interrupt)
+        Self::new(
+            ByteStreamSource::Read(Box::new(reader)),
+            span,
+            interrupt,
+            type_,
+        )
+    }
+
+    /// Create a [`ByteStream`] from a string. The type of the stream is always `String`.
+    pub fn read_string(string: String, span: Span, interrupt: Option<Arc<AtomicBool>>) -> Self {
+        let len = string.len();
+        ByteStream::read(
+            Cursor::new(string.into_bytes()),
+            span,
+            interrupt,
+            ByteStreamType::String,
+        )
+        .with_known_size(Some(len as u64))
     }
 
-    /// Create a new [`ByteStream`] from a [`ByteStreamSource::File`].
+    /// Create a [`ByteStream`] from a byte vector. The type of the stream is always `Binary`.
+    pub fn read_binary(bytes: Vec<u8>, span: Span, interrupt: Option<Arc<AtomicBool>>) -> Self {
+        let len = bytes.len();
+        ByteStream::read(Cursor::new(bytes), span, interrupt, ByteStreamType::Binary)
+            .with_known_size(Some(len as u64))
+    }
```

Comment:
I'm not sure. I thought they might be helpful if you want to write something that just works on ByteStream and want to adapt a plain value to it rather than the other way around.

Reply:
I guess adding these methods would allow the values to get the interrupt handling for free. Sure, let's give them a shot.